src/sqlfluff/dialects/dialect_ansi.py

```diff
@@ -803,7 +737,8 @@ class CaseExpressionSegment(BaseSegment):
             Ref('SelectStatementSegment'),
             Ref('LiteralGrammar'),
             Ref('IntervalExpressionSegment'),
-            Ref('ObjectReferenceSegment')
+            Ref('ObjectReferenceSegment'),
+            Ref('DatatypeIdentifierSegment')
```

Comment:
Here I think `DatatypeIdentifierSegment` should come before `ObjectReferenceSegment`, as 'ObjectReferenceSegment` will be greedy.